src/Mod/Sketcher/Gui/ViewProviderSketch.cpp

```diff
@@ -1210,17 +1210,129 @@ bool ViewProviderSketch::mouseButtonPressed(int Button, bool pressed, const SbVe
     }
     // Right mouse button ****************************************************
     else if (Button == 2) {
-        if (!pressed) {
+        if (pressed) {
+            // Do things depending on the mode of the user interaction
+            switch (Mode) {
+                case STATUS_NONE: {
+                    bool done = false;
+                    if (preselection.isPreselectPointValid()) {
+                        // Base::Console().Log("start dragging, point:%d\n",this->DragPoint);
+                        Mode = STATUS_SELECT_Point;
+                        done = true;
+                    }
+                    else if (preselection.isPreselectCurveValid()) {
+                        // Base::Console().Log("start dragging, point:%d\n",this->DragPoint);
+                        Mode = STATUS_SELECT_Edge;
+                        done = true;
+                    }
+                    else if (preselection.isCrossPreselected()) {
+                        // Base::Console().Log("start dragging, point:%d\n",this->DragPoint);
+                        Mode = STATUS_SELECT_Cross;
+                        done = true;
+                    }
+                    else if (!preselection.PreselectConstraintSet.empty()) {
+                        // Base::Console().Log("start dragging, point:%d\n",this->DragPoint);
+                        Mode = STATUS_SELECT_Constraint;
+                        done = true;
+                    }
+                }
+            }
+        }
+        else if (!pressed) {
             switch (Mode) {
                 case STATUS_SKETCH_UseHandler:
                     // delegate to handler whether to quit or do otherwise
                     sketchHandler->pressRightButton(Base::Vector2d(x, y));
                     return true;
                 case STATUS_NONE:
+                    generateContextMenu();
+                    return true;
                 case STATUS_SELECT_Point:
+                    if (pp) {
+                        // Base::Console().Log("Select Point:%d\n",this->DragPoint);
+                        //  Do selection
+                        std::stringstream ss;
+                        ss << "Vertex" << preselection.getPreselectionVertexIndex();
+
+                        if (!isSelected(ss.str())) {
+                            addSelection2(ss.str(),
+                                          pp->getPoint()[0],
+                                          pp->getPoint()[1],
+                                          pp->getPoint()[2]);
+                            drag.resetIds();
+                        }
```

Comment:
Can you refactor so we don't repeat this so many times?

Reply:
done, and also for the LMB event.